bot/src/bot.rs

```diff
@@ -0,0 +1,175 @@
+use std::{str::FromStr, sync::Arc};
+
+use anyhow::{anyhow, Context, Result};
+use bson::{doc, DateTime as BsonDateTime};
+use chrono::{Duration, Utc};
+use futures::TryStreamExt;
+use mongodb::options::FindOptions;
+use mongodb::Collection;
+use starknet::accounts::Account;
+use starknet::{
+    accounts::{Call, SingleOwnerAccount},
+    core::{
+        types::{BlockId, CallFunction, FieldElement},
+        utils::get_selector_from_name,
+    },
+    macros::selector,
+    providers::{Provider, SequencerGatewayProvider},
+    signers::LocalWallet,
+};
+use url::Url;
+
+use crate::discord::log_msg_and_send_to_discord;
+use crate::{
+    config::Config,
+    models::{AppState, Approval, AutoRenewals, Domain},
+};
+use bigdecimal::{BigDecimal, Zero};
+
+pub fn get_provider(config: &Config) -> SequencerGatewayProvider {
+    if config.devnet_provider.is_devnet {
+        SequencerGatewayProvider::new(
+            Url::from_str(&config.devnet_provider.gateway).unwrap(),
+            Url::from_str(&config.devnet_provider.feeder_gateway).unwrap(),
+        )
+    } else {
+        SequencerGatewayProvider::starknet_alpha_goerli()
+    }
+}
+
+pub async fn get_domains_ready_for_renewal(
+    config: &Config,
+    state: &Arc<AppState>,
+) -> Result<(Vec<FieldElement>, Vec<FieldElement>)> {
+    let domains = state.db.collection::<Domain>("domains");
+    let renewals = state.db.collection::<AutoRenewals>("auto_renewals");
+    let approvals = state.db.collection::<Approval>("approvals");
+
+    let min_expiry_date = Utc::now() + Duration::days(120);
+    let find_options = FindOptions::builder().build();
+    let mut domain_cursor = domains
+        .find(
+            doc! { "expiry": { "$lt": BsonDateTime::from_millis(min_expiry_date.timestamp_millis()) } },
+            find_options,
+        )
+        .await?;
+
+    let all_domains = domain_cursor.try_collect::<Vec<_>>().await?;
+    let futures: Vec<_> = all_domains
```

Comment:
Doing one query per domain won't scale very well, let's try to do a single query with an aggregate and a project.